src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs

```diff
@@ -0,0 +1,410 @@
+
+            /// <summary>
+            /// The abstract method that specifies the NA value for the dst type.
+            /// </summary>
+            /// <returns></returns>
+            protected abstract void SetNaOutput(ref TOutput dst);
+
+            /// <summary>
+            /// The abstract method that realizes the main logic for the transform.
+            /// </summary>
+            /// <param name="input">A reference to the input object.</param>
+            /// <param name="dst">A reference to the dst object.</param>
+            /// <param name="windowedBuffer">A reference to the windowed buffer.</param>
+            /// <param name="iteration">A long number that indicates the number of times TransformCore has been called so far (starting value = 0).</param>
+            protected abstract void TransformCore(ref TInput input, FixedSizeQueue<TInput> windowedBuffer, long iteration, ref TOutput dst);
+
+            /// <summary>
+            /// The abstract method that realizes the logic for initializing the state object.
+            /// </summary>
+            protected abstract void InitializeStateCore();
+
+            /// <summary>
+            /// The abstract method that realizes the logic for learning the parameters and the initial state object from data.
+            /// </summary>
+            /// <param name="data">A queue of data points used for training</param>
+            protected abstract void LearnStateFromDataCore(FixedSizeQueue<TInput> data);
+        }
+
+        protected readonly IHost Host;
+
+        /// <summary>
+        /// The window size for buffering.
+        /// </summary>
+        protected readonly int WindowSize;
+
+        /// <summary>
+        /// The number of datapoints from the beginning of the sequence that are used for learning the initial state.
+        /// </summary>
+        protected int InitialWindowSize;
+
+        public string InputColumnName;
+        public string OutputColumnName;
+        protected ColumnType OutputColumnType;
+
+        public bool IsRowToRowMapper => false;
+
+        /// <summary>
+        /// The main constructor for the sequential transform
+        /// </summary>
+        /// <param name="windowSize">The size of buffer used for windowed buffering.</param>
+        /// <param name="initialWindowSize">The number of datapoints picked from the beginning of the series for training the transform parameters if needed.</param>
+        /// <param name="inputColumnName">The name of the input column.</param>
+        /// <param name="outputColumnName">The name of the dst column.</param>
+        /// <param name="name"></param>
+        /// <param name="env">A reference to the environment variable.</param>
+        /// <param name="outputColType"></param>
+        protected SequentialTransformerBase(int windowSize, int initialWindowSize, string inputColumnName, string outputColumnName,
+            string name, IHostEnvironment env, ColumnType outputColType)
```

Comment:
>IHostEnvironment [](start = 25, length = 16)

This is always the first parameter in all our ctors. Also, the 'env plus name' pattern is uncommon: we tend to just take `IHost` in those protected ctors